docs/compiler_passes.rst

```diff
@@ -1,11 +1,150 @@
 
+Most of the passes in their function and implementation are platform independent,
+deriving their platform dependent information from options and/or the configuration file.
+This holds also for scheduling and mapping, although one wouldn't think so first.
+All passes like this are mentioned below and describe platform independently:
 
-- decomposition
-- scheduling
-- mapping
+- their API, including their name: in general, apart from their name, they take all parameters from the ``program`` context which includes the configuration file and the platform, the options, and the vector of kernels with their circuits
+- the Intermediate Representation (IR) they expect as input and what they update in the IR; in general, they should accept any IR, so all types of gates, quantum as well as classical
+- the particular options they listen to; there usually is an option to disable it; also there are ways to dump the IR before and/or after it although this is not generally possible yet.
+
+Other passes, of which the function and implementation is platform dependent, can be found with the platforms.
+An example of the latter passes is QISA (i.e. instruction) generation in CC-Light.
+
+Passes have some general facilities available to them; these are not passes themselves since they don't transform the IR.
+Such facilities are:
+
+- Writing the IR out in an external representation (QASM 1.0) to a file;
+	the name of the file should relate to the point in compilation relative to the called passes where
+	the IR is written
+
+- Writing a report out to a file containing a summary of the IR;
+	e.g. the number of kernels,
+	the numer of one-qubit, two-qubit and more-qubit gates,
+	which qubits were uses and which not,
+	the wall clock time that compilation took until this point,
+	etc.
+
+- Writing a string to a file for off-line inspection
+	such as writing (in dot format) the gate dependence graph which is a scheduling pass internal data structure.
+	The writing to a file of a string is a general facility
+	but the generation of the string representation of the internal data structure is pass dependent.
+
+These facilities are described elsewhere.
+
+Writing the IR out to a file in a form suitable for a particular subsequent tool such as quantumsim
+is considered code generation for the quantumsim platform and is therefore considered a pass.
+
+:Note: A compiler pass is not something defined in OpenQL. It should be. Passes then have a standard API, standard intermediate representation dumpers before and after them, a standard way to include them in the compiler. We could have the list of passes to call be something defined in the configuration file, perhaps with the places where we want to have dumps and reports.
+
+Compiler passes in OpenQL are the compiler elements that, when called one after the other,
+gradually transform the OpenQL input program to some platform defined output program.
+The following passes are available and usually called in this order.
```

Comment:
What does "usually called in this order" mean? Where is documented how this works?